labman/db/tests/test_process.py

```diff
@@ -988,12 +988,16 @@ def test_create(self):
 
         plates = obs.plates
         self.assertEqual(len(plates), 1)
-        layout = plates[0].layout
-        self.assertEqual(layout[0][0].composition.i5_composition,
-                         PrimerComposition(1523))
-        self.assertEqual(layout[0][0].composition.i7_composition,
-                         PrimerComposition(1524))
-        self.assertIsNone(layout[-1][-1])
+        obs_primer_layout = [[[well.well_id if well is not None else None,
```

Comment:
😒 this tests is kind of odd, right? Basically here we are generating obs_primer_layout within the test vs. in a single function that generates this result ... this makes me wonder if we should be testing like this or some other way ...

Reply:
> Basically here we are generating obs_primer_layout within the test vs. in a single function that generates this result

No, not quite! The big list comprehension here is *not* generating the layout, just *reading* the layout generated by `LibraryPrepShotgunProcess.create` into a convenience format.

When `LibraryPrepShotgunProcess.create` makes `obs`, it fills the`obs.plates[0].layout` property with a list of lists of lists of Wells.   This makes it very hard to create a test case: you have to know the database id of each of the wells to instantiate all the expected Well objects, and it is prohibitively time-consuming for the code to instantiate them in the test case and compare them to all the Well objects in`obs.plates[0].layout`. 

Because of this, I chose to set up the part of the test that checks whether the correct i5 and i7 primer has been assigned to the correct well using a known-good as a list of lists of lists of *strings*: the human-readable well id--A1, etc, the i7 primer name, and the i5 primer name.  The known-good, in this format, is stored in `SHOTGUN_PRIMER_LAYOUT`.  The list comprehension you see is simply **extracting** those strings from`obs.plates[0].layout` (into a convenience variable `obs_primer_layout`) so that they can be compared with the known-good via `assertListEqual`.

In the cases where a given Well object in `obs.plates[0].layout` is None, of course you can't access its various nested string properties, so None is returned instead of the strings.

If you have thoughts for how to adjust the list comprehension to make it more readable, that would be great!

Reply:
perhaps worth adding this explanation?

Reply:
Good idea--will do!